src/socket.rs

```diff
@@ -0,0 +1,383 @@
+
+        self.state_command_sender
+            .send(StateCommand::Connect(Connect {
+                created_at,
+                timeout,
+                connected_sender,
+            }))
+            .await?;
+
+        time::timeout_at(deadline, connected_receiver).await??
+    }
+
+    /// Disconnect the client, regardless of any outstanding channel references
+    ///
+    /// Connected channels will return `ChannelError::Closed` when next used.
+    ///
+    /// New channels will need to be obtained from this client after `connect` is
+    /// called again.
+    pub async fn disconnect(&self) -> Result<(), DisconnectError> {
+        let (disconnected_sender, disconnected_receiver) = oneshot::channel();
+
+        self.state_command_sender
+            .send(StateCommand::Disconnect {
+                disconnected_sender,
+            })
+            .await?;
+
+        disconnected_receiver.await.map_err(From::from)
+    }
+
+    /// Propagates panic from [Listener::listen]
+    pub async fn shutdown(self) -> Result<(), ShutdownError> {
+        self.state_command_sender
+            .send(StateCommand::Shutdown)
+            .await
+            // if already shut down that's OK
+            .ok();
+
+        self.listener_shutdown().await
+    }
+
+    /// Creates a new, unjoined Phoenix Channel
+    pub async fn channel(
+        self: &Arc<Self>,
+        topic: &str,
+        payload: Option<Payload>,
+    ) -> Result<Arc<Channel>, ChannelError> {
+        let (sender, receiver) = oneshot::channel();
+
+        match self
+            .channel_spawn_sender
+            .send(ChannelSpawn {
+                socket: self.clone(),
+                topic: topic.to_string(),
+                payload,
+                sender,
+            })
+            .await
+        {
+            Ok(()) => receiver.await.map(Arc::new).map_err(From::from),
+            Err(_) => Err(self.listener_shutdown().await.unwrap_err().into()),
+        }
+    }
+
+    pub(crate) async fn join(
+        &self,
+        topic: Topic,
+        join_reference: JoinReference,
+        payload: Arc<Payload>,
+        deadline: Instant,
+    ) -> Result<oneshot::Receiver<Result<JoinedChannelReceivers, JoinError>>, JoinError> {
+        let (joined_sender, joined_receiver) = oneshot::channel();
+
+        match self
+            .channel_state_command_sender
+            .send(ChannelStateCommand::Join(Join {
+                topic,
+                join_reference,
+                payload,
+                deadline,
+                joined_sender,
+            }))
+            .await
+        {
+            Ok(()) => Ok(joined_receiver),
+            Err(_) => Err(self.listener_shutdown().await.unwrap_err().into()),
+        }
+    }
+
+    pub(crate) async fn leave(
+        &self,
+        topic: Topic,
+        join_reference: JoinReference,
+    ) -> Result<oneshot::Receiver<Result<(), LeaveError>>, LeaveError> {
```

Comment:
If you await the receiver then this return type can be reduced to Result<(), LeaveError>

Reply:
Similar to above with `join` above, I don't think I can await the receiver in `Socket::leave` because it would block until the server responds and not allow the `channel::Listener` to update the externally visible `Status` and not allow the `Leave` to be interrupted with a `Join` to cancel it or a `Shutdown`.

Reply:
Ok seems fine to leave as is